scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/policy/license/LicenseSynchronization.java

```diff
@@ -0,0 +1,218 @@
+        initDBFile();
+        initDBMaps(true);
+    }
+
+    private void initDBFile() {
+        storeFileDirectory = new File(PASchedulerProperties.getAbsolutePath(PASchedulerProperties.LICENSE_SCHEDULING_POLICY_DATABASE.getValueAsString()));
+        if (!storeFileDirectory.exists()) {
+            boolean result = storeFileDirectory.mkdirs();
+            if (!result) {
+                throw new IllegalArgumentException("Could not create directory hierarchy for path " +
+                                                   storeFileDirectory);
+            }
+        } else if (!storeFileDirectory.isDirectory()) {
+            throw new IllegalArgumentException("Provided directory path exists and is not a directory " +
+                                               storeFileDirectory);
+        }
+        storeFile = new File(storeFileDirectory, STORE_FILE_NAME);
+    }
+
+    private void initDBMaps(boolean firstAttempt) {
+        close();
+        try {
+            recordManager = createRecordManager(storeFile.getCanonicalPath());
+            persistedJobsLicenses = recordManager.hashMap(JOBS_LICENSES);
+            persistedTasksLicenses = recordManager.hashMap(TASKS_LICENSES);
+            remainingTokens = recordManager.hashMap(REMAINING_TOKENS);
+
+            recordManager.commit();
+            if (!firstAttempt) {
+                logger.info("Loading of job database successful after clean.");
+            }
+
+        } catch (Throwable e) {
+            if (firstAttempt) {
+                logger.error("Error occurred when loading the database " + storeFile.getAbsolutePath() +
+                             ", now cleaning it and retrying.", e);
+                cleanDataBase();
+                initDBMaps(false);
+            } else {
+                close();
+                throw new IllegalStateException("Error when loading database (even after cleaning it): " +
+                                                storeFile.getAbsolutePath(), e);
+            }
+        }
+    }
+
+    void addSoftware(String software, int nbTokens) {
+        persistedJobsLicenses.put(software, new LinkedBlockingQueue<String>(nbTokens));
+        persistedTasksLicenses.put(software, new LinkedBlockingQueue<String>(nbTokens));
+        remainingTokens.put(software, nbTokens);
+    }
+
+    void persist() {
+        try {
+            recordManager.commit();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    LinkedBlockingQueue<String> getPersistedJobsLicense(String license) {
+        return persistedJobsLicenses.get(license);
+    }
+
+    boolean containsJobId(String license, String jobId) {
+        return persistedJobsLicenses.get(license).stream().anyMatch(jid -> jid.equals(jobId));
+    }
+
+    void addJobToLicense(String license, String jobId) {
+        persistedJobsLicenses.get(license).add(jobId);
+        remainingTokens.put(license, remainingTokens.get(license) - 1);
```

Comment:
What if `remainingTokens.get(license)` is already 0? it will go negative, is it fine?

Reply:
In LicenseSchedulingPolicy, we only call this function after ensuring we have enough tokens (>0) with or without releasing tokens (from finished tasks).